examples/mario/src/main/scala/Effects.scala

```
@@ -0,0 +1,60 @@
+package mario
+
+import org.scalajs.dom
+import org.scalajs.dom.{KeyboardEvent, document}
+import org.scalajs.dom.raw.{Event, HTMLAudioElement, TouchEvent}
+import scalm.{Cmd, Sub, Task}
+import scalm.Sub.ofTotalObservable
+
+object Effects {
+
+  val requestAnimationFrameSub: Sub[Double] = ofTotalObservable[Double](
+    "requestAnimation", { observer =>
+      var handle = 0
+      def loop(time: Double): Unit = {
+        observer.onNext(time)
+        handle = dom.window.requestAnimationFrame(loop)
+      }
+      handle = dom.window.requestAnimationFrame(loop)
+      () =>
+        dom.window.cancelAnimationFrame(handle)
+    }
+  )
+
+  def keyPressSub[M](keyCode: Int, msg: M): Sub[M] = ofTotalObservable[M](
+    s"keyDown$keyCode", { observer =>
+      dom.window.addEventListener("keydown", (keyEvent: KeyboardEvent) => {
+        if (keyEvent.keyCode == keyCode) observer.onNext(msg)
+      })
+      () =>
+        ()
+    }
+  )
+
+  def keyReleaseSub[M](keyCode: Int, msg: M): Sub[M] = ofTotalObservable[M](
+    s"keyUp$keyCode", { observer =>
+      dom.window.addEventListener("keyup", (keyEvent: KeyboardEvent) => {
+        if (keyEvent.keyCode == keyCode) observer.onNext(msg)
+      })
+      () =>
+        ()
+    }
+  )
+
+  object Cmd {
+    def playSound[Msg](url: String, msgBack: Msg): Cmd[Msg] =
+      Task
+        .RunObservable[Unit, Msg] { _ =>
+          {
+            val audio =
+              document.createElement("audio").asInstanceOf[HTMLAudioElement]
+            audio.src = url
+            audio.onloadeddata = (_: Event) => audio.play()
+            () =>
+              ()
+          }
+        }
+        .attempt(_ => msgBack)
+  }
```

Since you ignore the error you should make this a `Cmd[Nothing]` so that users don’t have to create a “`Void`” message type.

---

Thx. I suppose so I should use `perform` instead of `attemp` to be able to return `Cmd[Nothing]` right?

---

Both should support producing a `Cmd[Nothing]` (I haven’t tried though, let me know if you encounter any issue)